File: binary-search/search-for-a-range.py
class Solution:
    def binary_search(self, A, B, left = True):
        low = 0
        high = len(A) - 1
        result = -1

        while low <= high:
            mid = low + (high - low) // 2
            if A[mid] == B:
                result = mid
                if left:
                    high = mid - 1
                else:
                    low = mid + 1
            elif A[mid] < B: low = mid + 1
            else: high = mid - 1

        return result

    def searchRange(self, A, B):
        left = self.binary_search(A, B)
        right = self.binary_search(A, B, False)
        return [left, right]

File: binary-search/search-for-a-range.cpp
vector<int> Solution::searchRange(const vector<int> &A, int B) {
   int n = A.size();
   int l = 0, r = n-1;
   int L = -1, R = -1, m;

   while(l <= r){
       m = (l + r) / 2;
       if(A[m] <= B){
           l = m+1; 
           if(A[m] == B) R = m;
       }
       else r = m - 1;
   }

   l = 0; r = R;

   while(l <= r){
       m = (l + r) / 2;
       if(A[m] >= B){
           r = m - 1;
           if(A[m] == B) L = m;
       }
       else l = m + 1;
   }

   vector<int> v;
   v.push_back(L);
   v.push_back(R);
   return v;
}

